src/lib/tuning/greedy_selector.cpp

```diff
@@ -0,0 +1,157 @@
+#include "greedy_selector.hpp"
+
+#include <algorithm>
+#include <limits>
+#include <list>
+
+#include "utils/logging.hpp"
+
+namespace opossum {
+
+/*
+ * This subroutine scans the least desirable available choices until the
+ * accumulated reject_cost() is lower than a required value.
+ * If the thereby accumulated reject_desirability is not lower than a threshold
+ * an iterator to the choice one past the subsequence to reject is returned.
+ * Otherwise the end() iterator is returned, that the constraints are not satisfied
+ * by any continuous subsequence starting at sorted_choices.begin().
+ */
+const std::list<std::shared_ptr<TuningChoice>>::const_iterator sacrifice_choices(
+    const std::list<std::shared_ptr<TuningChoice>>& sorted_choices, float required_cost_delta,
+    float acceptible_desirability_delta = -std::numeric_limits<float>::infinity()) {
+  auto desirability_delta = 0.0f;
+  auto cost_delta = 0.0f;
+
+  auto choice = sorted_choices.cbegin();
+  while (choice != sorted_choices.cend()) {
+    if (cost_delta <= required_cost_delta) {
+      break;
+    }
+    desirability_delta += (*choice)->reject_desirability();
+    cost_delta += (*choice)->reject_cost();
+    ++choice;
+  }
+
+  if (cost_delta <= required_cost_delta && desirability_delta >= acceptible_desirability_delta) {
+    return choice;
+  }
+  return sorted_choices.cend();
+}
+
+std::vector<std::shared_ptr<TuningOperation>> GreedySelector::select(
+    const std::vector<std::shared_ptr<TuningChoice>>& choices, float cost_budget) {
+  std::vector<std::shared_ptr<TuningOperation>> operations;
+  operations.reserve(choices.size());
+  // Assumption: cost() >= 0 ==> accept_cost() >= 0 && current_cost() >= 0 && reject_cost() <= 0
+
+  // Accumulate absolute cost balance from currently chosen choices.
+  float cost_balance = 0.0f;
+  for (const auto& choice : choices) {
+    cost_balance += choice->current_cost();
+  }
+
+  // Desirability balance is always relative to current system state.
+  float desirability_balance = 0.0f;
+
+  // Sort choices:
+  //  1) by reject_cost() ascending
+  //  2) by accept_desirability() ascending stable
+  // The most desirable choice is at back() and among equaly desirable choices
+  // the least costly choices to accept are nearer back().
+  std::list<std::shared_ptr<TuningChoice>> sorted_choices(choices.cbegin(), choices.cend());
+
+  sorted_choices.sort([](std::shared_ptr<TuningChoice> lhs, std::shared_ptr<TuningChoice> rhs) {
+    return lhs->reject_cost() < rhs->reject_cost();
+  });
+  sorted_choices.sort([](std::shared_ptr<TuningChoice> lhs, std::shared_ptr<TuningChoice> rhs) {
+    return lhs->accept_desirability() < rhs->accept_desirability();
```

Comment:
I‘m offline and can‘t check if `set::sort` is stable. Even if it is, I wonder why you don‘t write your lambda in a way where you only need one pass

Reply:
You are right, we changed the lambda to complete the job in one `std::sort` pass.